# Amazon Seller Sales & Fulfillment Analytics Platform

## 1. Import Libraries

## 2. Load Dataset

## 3. Create Working Copy

## 4. Remove Unnecessary Columns

## 5. Fix Data Types

## 6. Standardize Text Values

## 7. Handle Missing Values

## 8. Check Duplicates

## 9. Final Validation

## 10. Export Clean Dataset

In [21]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)

In [22]:
df = pd.read_csv("../data/raw/amazon_sales.csv")

C:\Users\user\AppData\Local\Temp\ipykernel_26484\1342654603.py:1: DtypeWarning: Columns (23) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("../data/raw/amazon_sales.csv")


In [23]:
df_clean = df.copy()

In [24]:
df_clean.head()

,index,Order ID,Date,Status,Fulfilment,Sales Channel,ship-service-level,Style,SKU,Category,Size,ASIN,Courier Status,Qty,currency,Amount,ship-city,ship-state,ship-postal-code,ship-country,promotion-ids,B2B,fulfilled-by,Unnamed: 22
0,0,405-8078784-5731545,04-30-22,Cancelled,Merchant,Amazon.in,Standard,SET389,SET389-KR-NP-S,Set,S,B09KXVBD7Z,NaN,0,INR,647.62,MUMBAI,MAHARASHTRA,400081.0,IN,NaN,False,Easy Ship,NaN
1,1,171-9198151-1101146,04-30-22,Shipped - Delivered to Buyer,Merchant,Amazon.in,Standard,JNE3781,JNE3781-KR-XXXL,kurta,3XL,B09K3WFS32,Shipped,1,INR,406.00,BENGALURU,KARNATAKA,560085.0,IN,Amazon PLCC Free-Financing Universal Merchant ...,False,Easy Ship,NaN
2,2,404-0687676-7273146,04-30-22,Shipped,Amazon,Amazon.in,Expedited,JNE3371,JNE3371-KR-XL,kurta,XL,B07WV4JV4D,Shipped,1,INR,329.00,NAVI MUMBAI,MAHARASHTRA,410210.0,IN,IN Core Free Shipping 2015/04/08 23-48-5-108,True,NaN,NaN
3,3,403-9615377-8133951,04-30-22,Cancelled,Merchant,Amazon.in,Standard,J0341,J0341-DR-L,Western Dress,L,B099NRCT7B,NaN,0,INR,753.33,PUDUCHERRY,PUDUCHERRY,605008.0,IN,NaN,False,Easy Ship,NaN
4,4,407-1069790-7240320,04-30-22,Shipped,Amazon,Amazon.in,Expedited,JNE3671,JNE3671-TU-XXXL,Top,3XL,B098714BZP,Shipped,1,INR,574.00,CHENNAI,TAMIL NADU,600073.0,IN,NaN,False,NaN,NaN


In [25]:
df_clean.columns

Index(['index', 'Order ID', 'Date', 'Status', 'Fulfilment', 'Sales Channel ',
       'ship-service-level', 'Style', 'SKU', 'Category', 'Size', 'ASIN',
       'Courier Status', 'Qty', 'currency', 'Amount', 'ship-city',
       'ship-state', 'ship-postal-code', 'ship-country', 'promotion-ids',
       'B2B', 'fulfilled-by', 'Unnamed: 22'],
      dtype='object')

In [26]:
df_clean["Date"] = pd.to_datetime(
    df_clean["Date"],
    errors="coerce"
)

C:\Users\user\AppData\Local\Temp\ipykernel_26484\2941966541.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df_clean["Date"] = pd.to_datetime(


In [27]:
df_clean["Date"].dtype

dtype('<M8[ns]')

In [28]:
df_clean["Date"].isnull().sum()

np.int64(0)

In [29]:
df_clean.head()

,index,Order ID,Date,Status,Fulfilment,Sales Channel,ship-service-level,Style,SKU,Category,Size,ASIN,Courier Status,Qty,currency,Amount,ship-city,ship-state,ship-postal-code,ship-country,promotion-ids,B2B,fulfilled-by,Unnamed: 22
0,0,405-8078784-5731545,2022-04-30,Cancelled,Merchant,Amazon.in,Standard,SET389,SET389-KR-NP-S,Set,S,B09KXVBD7Z,NaN,0,INR,647.62,MUMBAI,MAHARASHTRA,400081.0,IN,NaN,False,Easy Ship,NaN
1,1,171-9198151-1101146,2022-04-30,Shipped - Delivered to Buyer,Merchant,Amazon.in,Standard,JNE3781,JNE3781-KR-XXXL,kurta,3XL,B09K3WFS32,Shipped,1,INR,406.00,BENGALURU,KARNATAKA,560085.0,IN,Amazon PLCC Free-Financing Universal Merchant ...,False,Easy Ship,NaN
2,2,404-0687676-7273146,2022-04-30,Shipped,Amazon,Amazon.in,Expedited,JNE3371,JNE3371-KR-XL,kurta,XL,B07WV4JV4D,Shipped,1,INR,329.00,NAVI MUMBAI,MAHARASHTRA,410210.0,IN,IN Core Free Shipping 2015/04/08 23-48-5-108,True,NaN,NaN
3,3,403-9615377-8133951,2022-04-30,Cancelled,Merchant,Amazon.in,Standard,J0341,J0341-DR-L,Western Dress,L,B099NRCT7B,NaN,0,INR,753.33,PUDUCHERRY,PUDUCHERRY,605008.0,IN,NaN,False,Easy Ship,NaN
4,4,407-1069790-7240320,2022-04-30,Shipped,Amazon,Amazon.in,Expedited,JNE3671,JNE3671-TU-XXXL,Top,3XL,B098714BZP,Shipped,1,INR,574.00,CHENNAI,TAMIL NADU,600073.0,IN,NaN,False,NaN,NaN


In [32]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 128975 entries, 0 to 128974
Data columns (total 24 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   index               128975 non-null  int64         
 1   Order ID            128975 non-null  object        
 2   Date                128975 non-null  datetime64[ns]
 3   Status              128975 non-null  object        
 4   Fulfilment          128975 non-null  object        
 5   Sales Channel       128975 non-null  object        
 6   ship-service-level  128975 non-null  object        
 7   Style               128975 non-null  object        
 8   SKU                 128975 non-null  object        
 9   Category            128975 non-null  object        
 10  Size                128975 non-null  object        
 11  ASIN                128975 non-null  object        
 12  Courier Status      122103 non-null  object        
 13  Qty                 128975 no

In [34]:
df_clean["Status"] = df_clean["Status"].str.title()

df_clean["Fulfilment"] = df_clean["Fulfilment"].str.title()

df_clean["Courier Status"] = df_clean["Courier Status"].str.title()

In [35]:
missing = (
    df_clean.isnull()
            .sum()
            .sort_values(ascending=False)
)

missing[missing > 0]

fulfilled-by        89698
promotion-ids       49153
Unnamed: 22         49050
currency             7795
Amount               7795
Courier Status       6872
ship-country           33
ship-city              33
ship-state             33
ship-postal-code       33
dtype: int64

Drop useless columns

In [36]:
df_clean.drop(
    columns=[
        "fulfilled-by",
        "promotion-ids",
        "Unnamed: 22"
    ],
    inplace=True
)

In [39]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 128975 entries, 0 to 128974
Data columns (total 21 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   index               128975 non-null  int64         
 1   Order ID            128975 non-null  object        
 2   Date                128975 non-null  datetime64[ns]
 3   Status              128975 non-null  object        
 4   Fulfilment          128975 non-null  object        
 5   Sales Channel       128975 non-null  object        
 6   ship-service-level  128975 non-null  object        
 7   Style               128975 non-null  object        
 8   SKU                 128975 non-null  object        
 9   Category            128975 non-null  object        
 10  Size                128975 non-null  object        
 11  ASIN                128975 non-null  object        
 12  Courier Status      122103 non-null  object        
 13  Qty                 128975 no

Handle Amount and Currency

In [37]:
df_clean[
    df_clean["Amount"].isna()
][["Amount","currency"]].head()

,Amount,currency
8,NaN,NaN
29,NaN,NaN
65,NaN,NaN
84,NaN,NaN
95,NaN,NaN


These rows have:

No sale amount
No currency
Usually correspond to cancelled or invalid transactions
Cannot contribute to revenue calculations

If we keep them:

Revenue calculations become inaccurate.
Average order value gets distorted.
Power BI visuals may show blanks.
SQL aggregations become messier.

Since only about 6% (7,795 rows) are affected, dropping them is a reasonable business decision.

In [40]:
df_clean = df_clean.dropna(subset=["Amount"])

In [41]:
df_clean["Amount"].isna().sum()

np.int64(0)

In [42]:
df_clean["currency"].isna().sum()

np.int64(0)

Checking Courier Status

In [44]:
df_clean[
    df_clean["Courier Status"].isna()
][["Status", "Courier Status"]].head(20)

,Status,Courier Status
0,Cancelled,NaN
3,Cancelled,NaN
23,Cancelled,NaN
83,Cancelled,NaN
178,Cancelled,NaN
226,Cancelled,NaN
227,Cancelled,NaN
251,Cancelled,NaN
270,Cancelled,NaN
324,Cancelled,NaN


Fill it with "Not Shipped"

Missing Courier Status values correspond to cancelled orders that were never shipped. They are replaced with "Not Shipped" to preserve business meaning and simplify downstream analysis.

In [45]:
df_clean["Courier Status"] = df_clean["Courier Status"].fillna("Not Shipped")

C:\Users\user\AppData\Local\Temp\ipykernel_26484\3253117709.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean["Courier Status"] = df_clean["Courier Status"].fillna("Not Shipped")


Handle the remaining 33 missing shipping values

In [46]:
df_clean["ship-country"] = df_clean["ship-country"].fillna("IN")

df_clean["ship-state"] = df_clean["ship-state"].fillna("Unknown")

df_clean["ship-city"] = df_clean["ship-city"].fillna("Unknown")

df_clean["ship-postal-code"] = (
    df_clean["ship-postal-code"]
        .fillna(df_clean["ship-postal-code"].mode()[0])
)

C:\Users\user\AppData\Local\Temp\ipykernel_26484\2776900061.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean["ship-country"] = df_clean["ship-country"].fillna("IN")
C:\Users\user\AppData\Local\Temp\ipykernel_26484\2776900061.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean["ship-state"] = df_clean["ship-state"].fillna("Unknown")
C:\Users\user\AppData\Local\Temp\ipykernel_26484\2776900061.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.

In [47]:
missing = (
    df_clean.isnull()
            .sum()
            .sort_values(ascending=False)
)

missing[missing > 0]

Series([], dtype: int64)

Standardize Text Values

In [49]:
text_cols = df_clean.select_dtypes(include="object").columns

for col in text_cols:
    df_clean[col] = df_clean[col].str.strip()

C:\Users\user\AppData\Local\Temp\ipykernel_26484\900330049.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean[col] = df_clean[col].str.strip()


In [50]:
df_clean["Status"] = df_clean["Status"].str.title()

df_clean["Fulfilment"] = df_clean["Fulfilment"].str.title()

df_clean["Courier Status"] = df_clean["Courier Status"].str.title()

C:\Users\user\AppData\Local\Temp\ipykernel_26484\2726993822.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean["Status"] = df_clean["Status"].str.title()
C:\Users\user\AppData\Local\Temp\ipykernel_26484\2726993822.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_clean["Fulfilment"] = df_clean["Fulfilment"].str.title()
C:\Users\user\AppData\Local\Temp\ipykernel_26484\2726993822.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row

Check Duplicate Rows

In [52]:
df_clean.duplicated().sum()

np.int64(0)

Final Validation

In [53]:
df_clean.shape

(121180, 21)

In [54]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 121180 entries, 0 to 128974
Data columns (total 21 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   index               121180 non-null  int64         
 1   Order ID            121180 non-null  object        
 2   Date                121180 non-null  datetime64[ns]
 3   Status              121180 non-null  object        
 4   Fulfilment          121180 non-null  object        
 5   Sales Channel       121180 non-null  object        
 6   ship-service-level  121180 non-null  object        
 7   Style               121180 non-null  object        
 8   SKU                 121180 non-null  object        
 9   Category            121180 non-null  object        
 10  Size                121180 non-null  object        
 11  ASIN                121180 non-null  object        
 12  Courier Status      121180 non-null  object        
 13  Qty                 121180 non-nul

In [55]:
df_clean.isnull().sum()

index                 0
Order ID              0
Date                  0
Status                0
Fulfilment            0
Sales Channel         0
ship-service-level    0
Style                 0
SKU                   0
Category              0
Size                  0
ASIN                  0
Courier Status        0
Qty                   0
currency              0
Amount                0
ship-city             0
ship-state            0
ship-postal-code      0
ship-country          0
B2B                   0
dtype: int64

In [56]:
df_clean.head()

,index,Order ID,Date,Status,Fulfilment,Sales Channel,ship-service-level,Style,SKU,Category,Size,ASIN,Courier Status,Qty,currency,Amount,ship-city,ship-state,ship-postal-code,ship-country,B2B
0,0,405-8078784-5731545,2022-04-30,Cancelled,Merchant,Amazon.in,Standard,SET389,SET389-KR-NP-S,Set,S,B09KXVBD7Z,Not Shipped,0,INR,647.62,MUMBAI,MAHARASHTRA,400081.0,IN,False
1,1,171-9198151-1101146,2022-04-30,Shipped - Delivered To Buyer,Merchant,Amazon.in,Standard,JNE3781,JNE3781-KR-XXXL,kurta,3XL,B09K3WFS32,Shipped,1,INR,406.00,BENGALURU,KARNATAKA,560085.0,IN,False
2,2,404-0687676-7273146,2022-04-30,Shipped,Amazon,Amazon.in,Expedited,JNE3371,JNE3371-KR-XL,kurta,XL,B07WV4JV4D,Shipped,1,INR,329.00,NAVI MUMBAI,MAHARASHTRA,410210.0,IN,True
3,3,403-9615377-8133951,2022-04-30,Cancelled,Merchant,Amazon.in,Standard,J0341,J0341-DR-L,Western Dress,L,B099NRCT7B,Not Shipped,0,INR,753.33,PUDUCHERRY,PUDUCHERRY,605008.0,IN,False
4,4,407-1069790-7240320,2022-04-30,Shipped,Amazon,Amazon.in,Expedited,JNE3671,JNE3671-TU-XXXL,Top,3XL,B098714BZP,Shipped,1,INR,574.00,CHENNAI,TAMIL NADU,600073.0,IN,False


Export Clean Dataset

In [57]:
df_clean.to_csv(
    "../data/cleaned/amazon_sales_cleaned.csv",
    index=False
)

In [58]:
import os

os.listdir("../data/cleaned")

['amazon_sales_cleaned.csv']